### DISCUSSION TITLE
Container with State/Data?

### DISCUSSION THREAD

**Discussion by MrSkwiggs** [hmlongco/Factory#194] [Q&A] [answered]
I want to implement a custom container which holds a reference to an `AuthToken`, so that I may build services which require said token.

Unfortunately, there seems to be no way to go around the `static let shared = MyContainer()` requirement, which renders my attempts impossible :( 

Would it be possible to provide a `static func shared<T>(_ constructorData: T) -> Self` overload or something which would allow me to do this? 

Or is there a better, `Factory`-ish way of providing this behaviour, that I'm not aware of?

**Comment by hmlongco** [comment, marked as answer]
SharedContainer requires at least one instance of a shared container. Doesn't mean you can't create multiple instances of them and  ignore the shared container.

Or you can simply derive from the parent protocol, ManagedContainer.

**Reply by MrSkwiggs** [reply]
So wait, what happens if I do:

```swift
final class MyContainer: SharedContainer {
  private let token: String

  // MARK: - Do not use
  static let shared = MyContainer(token: "")
  
  static func shared(token: String) -> MyContainer {
    .init(token: token)
  }

  var myService: Factory<MyService> {
    self { .init(token: token) }
  }
}
```

In the case where dependencies are injected etc. How does it know which static instance to use?

**Reply by MrSkwiggs** [reply]
Ah I guess reading the code docs for this protocol, `shared` is used exactly for this, so it would not work in this case. But using just `ManagedContainer` seems to do the trick just fine, thanks 👍

**Reply by hmlongco** [reply]
Didn't have the code sample, so thought this was a different problem. If you're using Injected, etc., you still couldn't do what you want since there would be no way to tell the wrapper to pass a value through to shared.

I'd probably do...
```swift
final public class MyContainer: SharedContainer {

    public  static let shared = MyContainer()
    public let manager = ContainerManager()

    private var token: Factory<String> {
        self { /* obtain token from wherever token comes from */  }.cached
    }

    public var myService: Factory<MyService> {
        self { .init(token: token()) }
    }

}
```
Or if token _must_ be passed in from the outside on initialization, change the private to internal and do something like...
```swift
// someplace early in initialization
MyContainer.shared.token.register { "SUPER SECRET TOKEN" }
```
